libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -670,21 +663,51 @@ AP_GPS::broadcast_first_configuration_failure_reason(void) const {
     }
 }
 
-void
-AP_GPS::_broadcast_gps_type(const char *type, uint8_t instance, int8_t baud_index)
+void AP_GPS::_detection_message(char *buffer, uint8_t buflen, uint8_t instance)
 {
-    char buffer[64];
+    const struct detect_state dstate = detect_state[instance];
+    const int8_t baud_index = dstate.last_baud;
+    AP_GPS_Backend *driver = drivers[instance];
+    if (driver == nullptr) {
+        // we really shouldn't have been called
+        buffer[0] = '\0';
+        return;
+    }
+
     if (baud_index >= 0) {
-        hal.util->snprintf(buffer, sizeof(buffer),
+        hal.util->snprintf(buffer, buflen,
                  "GPS %d: detected as %s at %d baud",
                  instance + 1,
-                 type,
+                 driver->name(),
                  _baudrates[baud_index]);
     } else {
-        hal.util->snprintf(buffer, sizeof(buffer),
+        hal.util->snprintf(buffer, buflen,
                  "GPS %d: detected as %s",
                  instance + 1,
-                 type);
+                 driver->name());
     }
+}
+
+
+void AP_GPS::_broadcast_gps_type(uint8_t instance)
+{
+    char buffer[64];
+    _detection_message(buffer, sizeof(buffer), instance);
     GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, buffer);
 }
+
+void AP_GPS::Write_DataFlash_Log_Startup_messages()
+{
+    if (_DataFlash == nullptr) {
+        return;
+    }
+
+    for (uint8_t instance=0; instance<GPS_MAX_INSTANCES; instance++) {
+        char buffer[64];
+        _detection_message(buffer, sizeof(buffer), instance);
+        if (drivers[instance] == NULL || state[instance].status == NO_GPS) {
```

Comment:
NULL -> nullptr

Also why is this if not the first instruction inside the for? Is there a reason to call `_detection_message` if `status == NO_GPS`? In the other case (if the driver is null) it won't do nothing.